third_party/terraform/utils/compute_instance_network_interface_helpers.go

```diff
@@ -0,0 +1,183 @@
+package google
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/errwrap"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	computeBeta "google.golang.org/api/compute/v0.beta"
+)
+
+type networkInterfaceHelper struct {
```

Comment:
I really like that you used a new type for this! That's a pattern that I feel like we forget sometimes that we can do, since we live so much in generated-code-world. Instead of a struct just containing functions though, it might make more sense to have these actually be methods on the networkInterfaceHelper type. Then the struct itself can just keep track of properties that are needed in each of the funcs.

In case it's helpful, the batching logic that we have at https://github.com/hashicorp/terraform-provider-google/blob/master/google/batcher.go is another example of a time when we actually created an object to help out with some stuff instead of just passing values around between functions.

Reply:
hmm not sure I agree. refactoring like this would expose some of the variables that are currently scoped and abstracted from the user

Reply:
Just a quick drive-by!

Have you thought about creating a struct with a bunch of unexported (lower-case) fields on it? While a user could become aware of + change these values, it wouldn't be the intent and users wouldn't have to pay attention to those values. With a struct of functions, we lose the ability to be able to create an Interface can could be used to create different types of NetworkInterface objects down the line (or mocks).

Reply:
Hey @danawillow, just pinging on this. Let me know if this refactor is preferable for you?

Reply:
It is, but I still think there's more that can be done here. One of the things that's standing out to me is the mix of resource-scoped fields (d, config, etc.) vs iteration-scoped ones (index) vs things that may change throughout the run (networkInterface, etc.) I think it would be best to try to get this struct as close as possible to having the exact amount of information it needs- no more, no less (in order to make it not have like 100 fields you probably won't be able to get it to literally exactly as much as it needs, but for example I don't think it needs all of `d`). Then things that change over time can become function arguments and return values.

Another thing I want to make sure you're aware of is interfaces: https://golang.org/doc/effective_go.html#generality. Since you mentioned being concerned about being able to use the struct values directly, you should be able to use an interface instead so that the call site only interacts with the exported methods.

Reply:
Updated to the following 
```
1. explicitly pass fields from (d, config) instead of passing to network interface helper
2. moved mutate operations out of network interface helper
3. changed to interface implementation so consumer does not alter object